troop/views.py

```diff
@@ -106,3 +131,81 @@ def form_invalid(self, form):
         response = super().form_invalid(form)
         response.status_code = 422  # Unprocessable Entity
         return response
+
+
+class NamiSearchView(OnlyTroopManagerMixin, generic.FormView):
+    form_class = NamiSearchForm
+    template_name = "troop/nami_search_form.html"
+
+    def nami(self):
+        if not settings.NAMI_USERNAME or not settings.NAMI_PASSWORD:
+            raise Exception("empty nami settings")
+
+        return Nami(
+            {
+                "username": settings.NAMI_USERNAME,
+                "password": settings.NAMI_PASSWORD,
+                "session_file": settings.NAMI_SESSION,
+            }
+        )
+
+    def form_valid(self, form):
+        kwargs = {"troop": self.request.troop.number}
+        data = {"nami": form.cleaned_data["nami"]}
+
+        try:
+            p = Participant.objects.filter(**data).get()
+
+            if p.troop_id != self.request.troop.pk:
+                messages.add_message(
+                    self.request,
+                    messages.ERROR,
+                    _("Participant already registered with another troop."),
+                )
+
+                response = super().form_invalid(form)
+                response.status_code = 422  # Unprocessable Entity
+                return response
+
+            messages.add_message(
+                self.request, messages.INFO, _("Participant already registered."),
+            )
+
+            kwargs["pk"] = p.id
+            self.success_url = reverse("troop:participant.edit", kwargs=kwargs)
+            return super().form_valid(form)
+        except Participant.DoesNotExist:
+            pass
+
+        try:
+            nami = self.nami()
+            member = nami.find_member(
+                data["nami"], grp_number=self.request.troop.number
+            )
+            data = nami.member_normalized(member)
+
+            messages.add_message(
+                self.request,
+                messages.SUCCESS,
+                _("NaMi search succeeded. Please complete the participant details."),
+            )
+        except:  # noqa: E722
+            messages.add_message(
+                self.request,
+                messages.WARNING,
+                _("NaMi search failed. Please complete the participant details."),
```

Comment:
Would be good to print out the exception message here, otherwise it's hard to know why the lookup failed.

Reply:
It now outputs the name of the class (I'm afraid that exposing the whole exception might lead to security issues). Is it better now?

Reply:
Better. But in case of a wrong username/password combination, `ResponseTypeError` is not very helpful. Could we make this even more explicit? Like raise a `WrongPassword` exception?

Reply:
Good point. It now raises an `AuthenticationError`

Reply:
:ok_hand: